Library/Homebrew/formula.rb

```diff
@@ -1934,26 +1943,29 @@ def stage
     active_spec.stage do |staging|
       @source_modified_time = active_spec.source_modified_time
       @buildpath = Pathname.pwd
-      env_home = buildpath/".brew_home"
-      mkdir_p env_home
+      @env_home = @buildpath/"../brew_home"
 
       stage_env = {
         HOMEBREW_PATH: nil,
       }
 
       unless ARGV.interactive?
-        stage_env[:HOME] = env_home
+        stage_env[:HOME] = @env_home
         stage_env[:_JAVA_OPTIONS] =
           "#{ENV["_JAVA_OPTIONS"]} -Duser.home=#{HOMEBREW_CACHE}/java_cache"
         stage_env[:CURL_HOME] = ENV["CURL_HOME"] || ENV["HOME"]
       end
 
-      setup_home env_home
+      setup_home @env_home
 
       ENV.clear_sensitive_environment!
 
       begin
         with_env(stage_env) do
+          if (HOMEBREW_TEMP/"brew_home").exist? &&
+            ENV["HOME"] == (HOMEBREW_TEMP/"brew_home").realpath.to_s
+            raise "HOME is #{ENV["HOME"]}; refusing to proceed!"
```

Comment:
this should be checked before creating the directory

Triggered with
```
iMac-TMP:homebrew-core joe$ ls /tmp/brew_home
ls: /tmp/brew_home: No such file or directory
iMac-TMP:homebrew-core joe$ brew  pull  https://github.com/Homebrew/homebrew-core/pull/29337
==> Fetching patch 
Patch: https://github.com/Homebrew/homebrew-core/pull/29337.patch
######################################################################## 100.0%
==> Applying patch
Applying: unp64 2.35
Warning: unp64 has a bottle: do you need to update it with --bottle?
==> Patch closes issue #29337
==> Patch changed:
 Formula/unp64.rb | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)
iMac-TMP:homebrew-core joe$ brew instal -dvs unp64
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/unp64.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/p7zip.rb
/usr/bin/sandbox-exec -f /tmp/homebrew20180622-1349-1f917cd.sb nice /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.3_2/bin/ruby -W0 -I /usr/local/Homebrew/Library/Homebrew:/usr/local/Homebrew/Library/Homebrew/cask/lib -- /usr/local/Homebrew/Library/Homebrew/build.rb /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/unp64.rb --verbose --debug
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FromPathLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/unp64.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/p7zip.rb
==> Downloading http://iancoog.altervista.org/C/unp64_235.7z
Already downloaded: /Users/joe/Library/Caches/Homebrew/unp64-2.35.7z
==> Verifying unp64-2.35.7z checksum
7zr x /Users/joe/Library/Caches/Homebrew/unp64-2.35.7z

7-Zip (a) [64] 16.02 : Copyright (c) 1999-2016 Igor Pavlov : 2016-05-21
p7zip Version 16.02 (locale=utf8,Utf16=on,HugeFiles=on,64 bits,8 CPUs x64)

Scanning the drive for archives:
1 file, 277047 bytes (271 KiB)               

Extracting archive: /Users/joe/Library/Caches/Homebrew/unp64-2.35.7z
--
Path = /Users/joe/Library/Caches/Homebrew/unp64-2.35.7z
Type = 7z
Physical Size = 277047
Headers Size = 4551
Method = LZMA2:16 BCJ
Solid = +
Blocks = 2

Everything is Ok

Folders: 4
Files: 242
Size:       1280491
Compressed: 277047
Error: HOME is /private/tmp/brew_home; refusing to proceed!
/usr/local/Homebrew/Library/Homebrew/formula.rb:1967:in `block (2 levels) in stage'
/usr/local/Homebrew/Library/Homebrew/utils.rb:557:in `with_env'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1964:in `block in stage'
/usr/local/Homebrew/Library/Homebrew/resource.rb:119:in `block in unpack'
/usr/local/Homebrew/Library/Homebrew/extend/fileutils.rb:13:in `block in mktemp'
/usr/local/Homebrew/Library/Homebrew/extend/fileutils.rb:73:in `block in run'
/usr/local/Homebrew/Library/Homebrew/extend/fileutils.rb:73:in `chdir'
/usr/local/Homebrew/Library/Homebrew/extend/fileutils.rb:73:in `run'
/usr/local/Homebrew/Library/Homebrew/extend/fileutils.rb:12:in `mktemp'
/usr/local/Homebrew/Library/Homebrew/resource.rb:114:in `unpack'
/usr/local/Homebrew/Library/Homebrew/resource.rb:92:in `stage'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1943:in `stage'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1116:in `brew'
/usr/local/Homebrew/Library/Homebrew/build.rb:113:in `block in install'
/usr/local/Homebrew/Library/Homebrew/utils.rb:557:in `with_env'
/usr/local/Homebrew/Library/Homebrew/build.rb:110:in `install'
/usr/local/Homebrew/Library/Homebrew/build.rb:191:in `<main>'
iMac-TMP:homebrew-core joe$ ls /tmp/brew_home
Library
```

Reply:
Both points addressed. The `download_strategy` changes in particular still need _a lot_ of rewriting and testing.